data_structures/doubly_linked_list.js

```diff
@@ -48,37 +71,218 @@ class Node {
-    constructor(){}
+    constructor(){
+        this.head = null
+        this.tail = null
+    }
     
     //add a new node with the given value to the tail of the list
-    add_to_tail(value){}
+    add_to_tail(value){
+        let new_node = new Node(value)
+
+        if(!this.tail){
+            this.head = new_node
+            this.tail = new_node
+            
+            return this
+        }
+
+        let current_tail = this.tail
+        new_node.prev = current_tail
+        current_tail.next = new_node
+        this.tail = new_node
+
+        return this
+    }
     
     //returns the value of the removed node or null if list is empty
-    remove_tail(){}
+    remove_tail(){
+        if(!this.tail){
+            return null
+        }
+
+        let tail = this.tail
+
+        if(this.head === this.tail){
+            this.head = null
+            this.tail = null
+
+            return tail.value
+        }
+
+        this.tail = tail.prev
+        this.tail.next = null
+
+        return tail.value
+    }
     
     //add a new node with the given value to the head of the list
-    add_to_head(value){}
+    add_to_head(value){
+        let new_node = new Node(value)
+
+        if(!this.head){
+            this.head = new_node
+            this.tail = new_node
+
+            return this
+        }
+
+        let current_head = this.head
+
+        this.head = new_node
+        new_node.next = current_head
+        current_head.prev = new_node
+
+        return this
+    }
     
     //returns the value of the removed node or null if list is empty
-    remove_head() {}
+    remove_head() {
+        if(!this.head){
+            return null
+        }
+
+        let head = this.head
+
+        if(this.head === this.tail){
+            this.head = null
+            this.tail = null
+            return head.value
+        }
+
+        this.head = head.next
+        this.head.prev = null
+        return head.value
+    }
     
     //returns the max value within the list
-    get_max(){}
+    get_max(){
+        if(!this.head){
+            return null
+        }
+        else if(this.head === this.tail){
+            return this.head.value
+        }
+
+        let maxNum = Number.MIN_VALUE
+        let curr_node = this.head;
+
+        while(curr_node !== null){
+            if(maxNum < curr_node.value){
+                maxNum = curr_node.value
+            }
+            
+            curr_node = curr_node.next
+        }
+
+        return maxNum
+    }
     
     //returns the length of the list
-    len(){}
+    len(){
+        let current_node = this.head
+        let length = 0
+
+        while(current_node !== null){
+            length += 1
+
+            current_node = current_node.next
+        }
+
+        return length
+    }
     
     //removes the input node from its current spot in the 
     // List and inserts it as the new head node of the List
-    move_to_front(){}
+    move_to_front(node){
```

Comment:
While this works and is a way to do it you could have implemented the class functions of the linked list and node here like:

```javascript
if(node === this.head){
            return null;
        }
        
        let value = node.value;
        
        if(node === this.tail){
            this.remove_tail()
        } else {
            node.delete();
            this.length -= 1;
        }
        
        this.add_to_head(value)
```